opcua.cpp

```diff
@@ -1503,6 +1602,253 @@ SOPC_ReturnStatus OPCUA::createS2Subscription()
+			{
+				dv->Value.Value.Boolean = (SOPC_Boolean)u8;
+			}
+		}
+		else
+		{
+			scanRes = 0;
+		}
+		break;
+	case SOPC_SByte_Id:
+		scanRes = sscanf(val, "%d", &i8);
+		if (scanRes != 0 && i8 <= INT8_MAX && i8 >= INT8_MIN)
+		{
+			dv->Value.Value.Sbyte = (SOPC_SByte)i8;
+		}
+		else
+		{
+			scanRes = 0;
+		}
+		break;
+	case SOPC_Int16_Id:
+		scanRes = sscanf(val, "%" SCNi16, &dv->Value.Value.Int16);
+		break;
+	case SOPC_UInt16_Id:
+		scanRes = sscanf(val, "%" SCNu16, &dv->Value.Value.Uint16);
+		break;
+	case SOPC_Int32_Id:
+		scanRes = sscanf(val, "%" SCNi32, &dv->Value.Value.Int32);
+		break;
+	case SOPC_UInt32_Id:
+		scanRes = sscanf(val, "%" SCNu32, &dv->Value.Value.Uint32);
+		break;
+	case SOPC_Int64_Id:
+		scanRes = sscanf(val, "%" SCNi64, &dv->Value.Value.Int64);
+		break;
+	case SOPC_UInt64_Id:
+		scanRes = sscanf(val, "%" SCNu64, &dv->Value.Value.Uint64);
+		break;
+	case SOPC_Float_Id:
+		scanRes = sscanf(val, "%f", &dv->Value.Value.Floatv);
+		break;
+	case SOPC_Double_Id:
+		scanRes = sscanf(val, "%lf", &dv->Value.Value.Doublev);
+		break;
+	case SOPC_String_Id:
+		SOPC_String_Initialize(&dv->Value.Value.String);
+		if (SOPC_STATUS_OK == SOPC_String_CopyFromCString(&dv->Value.Value.String, val))
+		{
+			scanRes = 1;
+		}
+		break;
+	default:
+		Logger::getLogger()->error("Unsupported BuiltinId: %d", builtinId);
+		scanRes = 0;
+		break;
+	}
+
+	// Set the source timestamp for the value
+	dv->SourceTimestamp = SOPC_Time_GetCurrentTimeUTC();
+
+	// If parsing failed, clear and free the allocated memory
+	if (scanRes == 0)
+	{
+		Logger::getLogger()->error("Failed to parse value '%s' for BuiltinId: %d", val, builtinId);
+		SOPC_DataValue_Clear(dv);
+		SOPC_Free(dv);
+		return NULL;
+	}
+
+	return dv;
+}
+
+/**
+ * Reads the value of a given node and retrieves its BuiltinId and ArrayType.
+ *
+ * @param nodeIdStr          The node ID string of the node to read.
+ * @param outBuiltinTypeId   Pointer to store the BuiltinId of the node's value.
+ * @param outArrayType       Pointer to store the ArrayType of the node's value.
+ * @return                   True if the read operation is successful, false otherwise.
+ */
+bool OPCUA::read(const char *nodeIdStr, SOPC_BuiltinId *outBuiltinTypeId, SOPC_VariantArrayType *outArrayType)
+{
+	if (nodeIdStr == NULL || outBuiltinTypeId == NULL || outArrayType == NULL)
+	{
+		Logger::getLogger()->error("Invalid input parameters for read operation.");
+		return false;
+	}
+
+	SOPC_ReturnStatus status = SOPC_STATUS_NOK;
+	OpcUa_ReadRequest *readRequest = SOPC_ReadRequest_Create(1, OpcUa_TimestampsToReturn_Both);
+	OpcUa_ReadResponse *readResponse = NULL;
+
+	if (readRequest == NULL)
+	{
+		Logger::getLogger()->error("Failed to create ReadRequest.");
+		return false;
+	}
+
+	status = SOPC_ReadRequest_SetReadValueFromStrings(readRequest, 0, nodeIdStr, SOPC_AttributeId_Value, NULL);
+	if (status != SOPC_STATUS_OK)
+	{
+		Logger::getLogger()->error("Failed to set read value for node %s.", nodeIdStr);
```

Comment:
Can you add something that says why? That way the user might be able to act on the error